WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -9,46 +9,176 @@ import Alamofire
+            switch self {
+            case .default:
+                return nil
+            case let .reAuthentication(accountID):
+                return (try? context.existingObject(with: accountID))?.email
+            case let .jetpackSite(email):
+                return email
+            }
+        }
+    }
+
+    enum AuthenticationError: Error {
         case invalidCallbackURL
         case loginDenied(message: String)
         case obtainAccessToken
         case urlError(URLError)
         case parsing(DecodingError)
         case cancelled
-        case unknown(Swift.Error)
+        case unknown(Error)
+    }
+
+    enum SignInError: Error {
+        case authentication(AuthenticationError)
+        case fetchUser(Error)
+        case mismatchedEmail(expectedEmail: String)
+        case alreadySignedIn(signedInAccountEmail: String)
+        case loadingSites(Error)
+    }
+
+    private let coreDataStack: CoreDataStackSwift
+    private let authenticator: ((URL) throws(AuthenticationError) -> URL)?
+
+    init(
+        coreDataStack: CoreDataStackSwift = ContextManager.shared,
+        authenticator: ((URL) throws(AuthenticationError) -> URL)? = nil
+    ) {
+        self.coreDataStack = coreDataStack
+        self.authenticator = authenticator
     }
 
+    /// Sign in WP.com account.
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
     @MainActor
-    func signIn(from viewController: UINavigationController) async {
+    func signIn(from viewController: UIViewController, context: SignInContext) async -> TaggedManagedObjectID<WPAccount>? {
+        WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "start"])
+        do {
+            let account = try await attemptSignIn(from: viewController, context: context)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "success"])
+            return account
+        } catch {
+            present(error, from: viewController)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "error", "error": "\(error)"])
+            return nil
+        }
+    }
+
+    /// Sign in WP.com account.
+    ///
+    /// - SeeAlso: `signIn`
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
+    @MainActor
+    func attemptSignIn(from viewController: UIViewController, context: SignInContext) async throws(SignInError) -> TaggedManagedObjectID<WPAccount> {
+        let defaultAccount = try? WPAccount.lookupDefaultWordPressComAccount(in: coreDataStack.mainContext)
+        let hasAlreadySignedIn = defaultAccount != nil
+
         let token: String
         do {
-            token = try await authenticate(from: viewController)
+            token = try await authenticate(from: viewController, prefersEphemeralWebBrowserSession: hasAlreadySignedIn)
```

Comment:
What's the purpose of `prefersEphemeralWebBrowserSession`?

Reply:
I added it a while ago before disabling signing into multiple WP.com accounts. This can be useful if the app is already signed in, which most of the time means the Safari session is also signed in with the same WP.com user. `prefersEphemeralWebBrowserSession` is set to `true` at that scenario so that the user doesn't have to log out of the Safari session to sign in with another account.